File: src/org/newtonpolyhedron/solverprinters/UnimodularMatrixMakerPrinter.java
package org.newtonpolyhedron.solverprinters;

import java.io.PrintWriter;

import org.apache.commons.math3.fraction.BigFraction;
import org.apache.commons.math3.linear.FieldMatrix;
import org.fs.utils.collection.table.ArrayListTable;
import org.fs.utils.collection.table.Table;
import org.newtonpolyhedron.entity.SolverPrinter;
import org.newtonpolyhedron.solve.matrixuni.UnimodularMatrixMaker;
import org.newtonpolyhedron.utils.MatrixUtils;

public class UnimodularMatrixMakerPrinter extends SolverPrinter <UnimodularMatrixMaker> {
	
	private final FieldMatrix <BigFraction>	baseMatrix;
	
	public UnimodularMatrixMakerPrinter(
			final UnimodularMatrixMaker solver,
			final FieldMatrix <BigFraction> baseMatrix,
			final PrintWriter output) {
		super(solver, output);
		this.baseMatrix = baseMatrix;
	}
	
	@Override
	protected void solveFor(final UnimodularMatrixMaker solver, final PrintWriter output)
			throws Exception {
		final FieldMatrix <BigFraction> alpha = solver.getUnimodularFrom(baseMatrix);
		output.println();
		output.println("Alpha-matrix: ");
		output.println(MatrixUtils.toString(alpha));
	}
}

File: src/main/scala/org/newtonpolyhedron/solverprinters/UnimodularMatrixMakerPrinter.scala
package org.newtonpolyhedron.solverprinters

import java.io.PrintWriter

import org.newtonpolyhedron.NewtonImports._
import org.newtonpolyhedron.entity.SolverPrinter
import org.newtonpolyhedron.solve.matrixuni.UnimodularMatrixMaker
import org.newtonpolyhedron.utils.StringUtils

class UnimodularMatrixMakerPrinter[N <: MPNumber](
  override val solver: UnimodularMatrixMaker[N],
  val baseMatrix:      Matrix[N],
  override val output: PrintWriter
)(implicit mp: MathProcessor[N])
    extends SolverPrinter[UnimodularMatrixMaker[N]](solver, output) {

  override def solveFor(
      solver: UnimodularMatrixMaker[N],
      output: PrintWriter
  ) = {
    val alpha = solver.unimodularFrom(baseMatrix)
    val alphaInv = alpha.inverse
    // Using .toString as a workaround for https://github.com/scala/bug/issues/10573
    output.println(title("""Unimodular "Alpha" matrix"""))
    val text1 = new StringBuilder()
    text1 ++= subheader("Base matrix:") + "\n"
    text1 ++= baseMatrix.toString + "\n"
    val text2 = new StringBuilder()
    text2 ++= subheader("Alpha-matrix:") + "\n"
    text2 ++= alpha.toString + "\n"
    val text3 = new StringBuilder()
    text3 ++= subheader("Inverse alpha-matrix:") + "\n"
    text3 ++= alphaInv.toString + "\n"
    output.println(StringUtils.appendToRight(5, text1, text2, text3))
  }
}
